1-Draft/RFCNNNN-Dispose-Script-Block.md

```diff
@@ -0,0 +1,242 @@
+
+### Execution
+
+The `dispose{}` block will execute:
+
+1. After `end{}` during normal command operation, and
+2. During command disposal that occurs when a terminating exception is thrown during pipeline operations.
+    - If the dispose block has already been called (as it is after the `end{}` step), it will simply return immediately.
+
+### Behaviour
+
+The `dispose{}` block silently **discards** anything submitted to the success output stream from within it.
+
+Documentation should advise that `dispose{}` is _not intended_ for any standard output and will completely ignore it.
+Additionally, all other PowerShell data streams are accessible from `dispose{}`, meaning there is no issue sending data using any of the following streams from within `dispose{}`:
+
+- Error
+- Warning
+- Verbose
+- Debug
+- Information
+
+#### Error States
+
+Terminating errors can also be thrown from a `dispose{}` block, which will skip the invocation of the rest of that `dispose{}` block, without preventing other `dispose{}` blocks from being executed.
+
+### Implementation Details
+
+- A new `dispose` keyword, indicating a `NamedBlockAst` similarly to `begin`, `process`, and `end`, which is valid in the same contexts as the existing named block keywords.
+  - This includes new token types and appropriate entries in the `Tokenizer`'s keyword tables.
+- Additional members and constructors for `ScriptBlockAst` to enable it to be recognised by the parser.
+- Additional methods in `DlrScriptCommandProcessor` and `CompiledScriptBlock` to enable the block to be executed at the correct time.
+  - Both types implement or inherit from a type that implements `IDisposable`, and as such the existing `Dispose()` methods can generally be utilised to run this block.
+- Changes to the behaviour of command execution in `CommandProcessorBase` to recognise when a function or script cmdlet is being executed, and appropriately call their `Dispose()` methods to invoke the dispose block.
+  - This can include changes to the existing command disposal behaviour, to enable commands to dispose their resources as soon as their tasks in the pipeline are completed, instead of waiting for the entire pipeline to complete.
+  - Also required with this is an additional layer of `try/finally` during `Dispose()` events as we are opening up the possibility for users to throw terminating errors _during_ a command's disposal, and as such we must ensure that disposal still completes appropriately.
+
+#### `Cmdlet` / `PSCmdlet`
+
+Commands inheriting from `Cmdlet` or `PSCmdlet` can already implement a version of this behaviour.
+To do so, such commands need also implement `IDisposable` and the `Dispose()` method.
+This RFC would simply change when the dispose method is called.
+Currently, the method is called during pipeline disposal.
+This RFC would ensure it is called immediately after `EndProcessing()` completes during normal pipeline operation.
```

Comment:
Good question. Strictly speaking, it doesn't _need_ to be changed. 

However, I think it makes more sense to call it after `EndProcessing()` when possible, simply because that means that we:

1. have a more well-defined point where we can expect Dispose() to be called,
1. free up resources earlier, and
1. spread out the overall impact of calling `Dispose()` so that some of it can happen during pipeline execution / towards the end of a pipeline, rather than all waiting till the very end of a pipeline and having to wait for every command in the pipeline to be disposed, one at a time.

On paper I don't think the differences will be massively different in the majority of cases. Selfishly, this makes it slightly easier to implement in `CommandProcessorBase`, if I recall that section of code correctly. 😉

Reply:
Looking further at the code behaviour... I don't think this is strictly necessary. Although some work is currently done before disposing commands, it doesn't look like there's a significant wait between calling the endprocessing/DoComplete methods and the DisposeCommands(), so it's likely okay to leave it more or less where it is.

Reply:
Having been tinkering a bit more with this, it really comes down to a fairly small-ish difference... in most cases. 

- If we have `dispose` execute after _all_ `EndProcessing()`/`end{}` steps are completed, that is a bit more consistent with how other steps work (all cmdlets execute `begin` before moving to `process` for each, etc.)
- If we have `dispose` execute after _each_ `EndProcessing()`/`end{}` step is completed, that lets us free up resources a bit quicker, and makes some types of pipelines easier to work with. For example, `Get-Content $file | Some-Function | Set-Content $file` would be possible if `Get-Content` opens its file read handle in `EndProcessing()` and then disposes of it in `Dispose()`; `Set-Content` would freely be able to open its own file handle by the time its `EndProcessing()` step comes around. 
  - Currently, this requires parentheses to be used to separate the pipelines: `(Get-Content $file) | Some-Function | Set-Content $file`.

That's just an example, but I'm sure there are other potential applications that are a bit harder to reason about. I tend to think it's a better idea to dispose a bit sooner if we can.